### PR TITLE
Personal Runechat Color

### PR DESCRIPTION
**Pull request by XSlayer300**
## About The Pull Request
As I requested in #feeback, this is a port of _that one specific server_ that allows us snowflakes to choose our own Runechat color. I also did improvements on the mob_helpers code as well, so it can be a bit of a refractor.

image

image

## Why It's Good For The Game
Again, snowflakes.

## Changelog
:cl:
add: Runechat colors can either be randomized or based on your own preferences!
/:cl:

### CODE REVIEW

**Comment by tgstation-server** [comment]
You currently have a negative Fix/Feature pull request delta of -27. Maintainers may close this PR at will. Fixing issues or improving the codebase will improve this score.

File: code/datums/chatmessage.dm [hunk 1]
```diff
@@ -109,6 +109,23 @@
...
+			target.chat_color = GLOB.runechat_color_names[target.name]
+		else if (ismob(target) && M.client?.prefs?.enable_personal_chat_color && M.name == M.real_name && M.name == M.client.prefs.real_name)
+			var/per_color = M.client.prefs.personal_chat_color
+			GLOB.runechat_color_names[target.name] = per_color
+			target.chat_color = per_color
+		else
+			target.chat_color = colorize_string(target.name)
+
+		target.chat_color_darkened = color_shift(target.chat_color, 0.85, 0.85)
+		target.chat_color_name = target.name
```
Comment (TiviPlus) [inline code/datums/chatmessage.dm:126]: Why is most of this junk not in Login() or pref creation

File: code/datums/chatmessage.dm [hunk 2]
```diff
@@ -270,16 +287,27 @@
...
+			final_val = "#[num2hex(x, 2)][num2hex(m, 2)][num2hex(c, 2)]"
 		if(5)
-			return "#[num2hex(c, 2)][num2hex(m, 2)][num2hex(x, 2)]"
+			final_val = "#[num2hex(c, 2)][num2hex(m, 2)][num2hex(x, 2)]"
+
+	GLOB.runechat_color_names[name] = final_val
+	return final_val
+
+/datum/chatmessage/proc/color_shift(color, sat_shift = 1, lum_shift = 1)
+	var/list/HSL = rgb2hsl(hex2num(copytext(color, 2, 4)), hex2num(copytext(color, 4, 6)), hex2num(copytext(color, 6, 8)))
```
Comment (TiviPlus) [inline code/datums/chatmessage.dm:309]: All new procs must be documented

File: code/__DEFINES/runechat_defines.dm [hunk 3]
```diff
@@ -0,0 +1 @@
+GLOBAL_LIST_EMPTY(runechat_color_names)
```
Comment (TiviPlus) [inline code/__DEFINES/runechat_defines.dm:1]: Why is this global needed exactly

File: code/modules/client/preferences.dm [hunk 4]
```diff
@@ -1003,13 +1006,27 @@ GLOBAL_LIST_EMPTY(preferences_datums)
...
+		if("personal_chat_color")
+			var/new_chat_color = input(user, "Choose your character's runechat color:", "Character Preference",personal_chat_color) as color|null
+			if(new_chat_color)
+				var/list/temp_hsl = rgb2hsl(ReadRGB(new_chat_color)[1],ReadRGB(new_chat_color)[2],ReadRGB(new_chat_color)[3])
+				if(new_chat_color == "#000000")
+					personal_chat_color = "#FFFFFF"
+				else if(temp_hsl[3] >= 0.65 && temp_hsl[2] >= 0.15)
+					personal_chat_color = sanitize_hexcolor(new_chat_color, 6, 1)
+				else
+					to_chat(user, "<span class='danger'>Invalid color. Your color is not bright enough.</span>")
```
Comment (TiviPlus) [inline code/modules/client/preferences.dm:1028]: early return

File: code/modules/client/preferences_savefile.dm [hunk 5]
```diff
@@ -109,6 +109,9 @@
 			save_character()
 			return FALSE
 
+	S["enable_personal_chat_color"]			>> enable_personal_chat_color
+	S["personal_chat_color"]			>> personal_chat_color
+
```
Comment (TiviPlus) [inline code/modules/client/preferences_savefile.dm:114]: You dont read properly from savefile here?

File: code/modules/mob/mob_helpers.dm [hunk 6]
```diff
@@ -13,8 +13,70 @@ proc/is_blind(A)
 proc/hasorgans(A)
 	return ishuman(A)
 
-/proc/hsl2rgb(h, s, l)
-	return //TODO: Implement
+//colour formats
+/proc/rgb2hsl(red, green, blue)
```
Comment (TiviPlus) [inline code/modules/mob/mob_helpers.dm:17]: Why is this in mob helpers of all places
Also new procs must be documented

File: code/datums/chatmessage.dm [hunk 7]
```diff
@@ -270,16 +287,27 @@
...
+			final_val = "#[num2hex(m, 2)][num2hex(x, 2)][num2hex(c, 2)]"
 		if(4)
-			return "#[num2hex(x, 2)][num2hex(m, 2)][num2hex(c, 2)]"
+			final_val = "#[num2hex(x, 2)][num2hex(m, 2)][num2hex(c, 2)]"
 		if(5)
-			return "#[num2hex(c, 2)][num2hex(m, 2)][num2hex(x, 2)]"
+			final_val = "#[num2hex(c, 2)][num2hex(m, 2)][num2hex(x, 2)]"
+
+	GLOB.runechat_color_names[name] = final_val
+	return final_val
```
Comment (TiviPlus) [inline code/datums/chatmessage.dm:306]: Why is this glob still here?

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Comment by TiviPlus** [comment]
Author has expressed that they dont believe to have the code knowledge to complete this PR but they are free to ask for a reopen anytime

**Pull request closed without merging**